python/cudf/cudf/core/column/decimal.py

```diff
@@ -364,18 +364,51 @@ def _get_decimal_type(lhs_dtype, rhs_dtype, op):
     else:
         raise NotImplementedError()
 
+    if isinstance(lhs_dtype, type(rhs_dtype)):
+        # SCENARIO 1: If `lhs_dtype` & `rhs_dtype` are same, then try to
+        # see if `precision` & `scale` can be fit into this type.
+        try:
+            return lhs_dtype.__class__(precision=precision, scale=scale)
+        except ValueError:
+            # Call to _validate fails, which means we need
+            # to goto SCENARIO 3.
+            pass
+    else:
+        # SCENARIO 2: If `lhs_dtype` & `rhs_dtype` are of different dtypes,
+        # then try to see if `precision` & `scale` can be fit into the type
+        # with greater MAX_PRECISION (i.e., the bigger dtype).
+        try:
+            if lhs_dtype.MAX_PRECISION > rhs_dtype.MAX_PRECISION:
+                return lhs_dtype.__class__(precision=precision, scale=scale)
+            else:
+                return rhs_dtype.__class__(precision=precision, scale=scale)
+        except ValueError:
+            # Call to _validate fails, which means we need
+            # to goto SCENARIO 3.
+            pass
+
+    # SCENARIO 3: If either of the above two scenarios fail, then get the
+    # MAX_PRECISION of `lhs_dtype` & `rhs_dtype` so that we can only check
+    # and return a dtype that is greater than or equal to input dtype that
+    # can fit `precision` & `scale`.
+    lhs_rhs_max_precision = max(
+        lhs_dtype.MAX_PRECISION, rhs_dtype.MAX_PRECISION
+    )
     for decimal_type in (
         cudf.Decimal32Dtype,
         cudf.Decimal64Dtype,
         cudf.Decimal128Dtype,
     ):
-        try:
-            min_decimal_type = decimal_type(precision=precision, scale=scale)
-        except ValueError:
-            # Call to _validate fails, which means we need
-            # to try the next dtype
-            pass
-        else:
-            return min_decimal_type
+        if decimal_type.MAX_PRECISION >= lhs_rhs_max_precision:
```

Comment:
It looks like this `if` is checking the same thing as the `_validate` method in the decimal dtype constructor. Is this unnecessarily duplicated? I'd fall back on the `try` and remove the `if` if possible.

Reply:
I might be wrong here -- I see you're constructing the returned type with `precision=precision` instead of `precision=max_precision`. Would it be better to try and construct a type with `max_precision` but return a type with `precision` if that succeeds?

Reply:
> I might be wrong here -- I see you're constructing the returned type with `precision=precision` instead of `precision=max_precision`. Would it be better to try and construct a type with `max_precision` and return a type with `precision` if that succeeds? (Or is that a bug -- should it be returning a type with `max_precision`?)

It's not a bug, the dtype is expected to have `precision`. 


> It looks like this `if` is checking the same thing as the `_validate` method in the decimal dtype constructor. Is this unnecessarily duplicated? I'd fall back on the `try` and remove the `if` if possible.

This was a necessary duplication because we want to pick a `dtype` that is not less than `lhs_dtype` or `rhs_dtype`.